### ISSUE TITLE
Better multi-line string literals + deprecate `QuasiQuotes`

### ISSUE THREAD

**Issue by Ericson2314** [ghc-proposals/ghc-proposals#260]
It came up in https://github.com/ghc-proposals/ghc-proposals/pull/125 that `QuasiQuotes` is basically a work-around for not have good multi-line string literal syntax (in conjunction with TH splices), and furthermore one that has a number of downsides.

Over the long term I think it would be good to deprecate it for those reasons. Deprecating features in wide use seems scary but 1) we can leave it around for a while after deprecating, 2) we could have tools to automatically rewrite code to its replacement. Opening this issue in hopes that someone writes a proposal for this.

**Comment by simonmar** [comment]
Just as a remark, people really do use QuasiQuotation today as a more convenient syntax for multiline string literals. The idea has been independently discovered multiple times and there are (at least) 3 packages for it:

* http://hackage.haskell.org/package/raw-strings-qq
* http://hackage.haskell.org/package/heredoc
* http://hackage.haskell.org/package/heredocs

and there have been previous proposals for here document syntax, a quick search turned up https://mail.haskell.org/pipermail/haskell-prime/2009-April/002784.html but there might also be more.

**Comment by 23Skidoo** [comment]
A major problem with the existing multi-line string literal syntax is that it conflicts with CPP. I believe that `cpphs` solves this, however, GHC doesn't use it by default.

I'm +1 on adding built-in Python-like """-literals and raw strings.

**Comment by steshaw** [comment]
The [neat-interpolation](https://hackage.haskell.org/package/neat-interpolation) package handles indentation nicely.

**Comment by gridaphobe** [comment]
The multi-line string literals could certainly be improved, but I don't think we should remove QuasiQuotes. QuasiQuotes are user-defined syntactic extensions to source Haskell; I think replacing them with string literals + splices would obscure that purpose.

**Comment by Ericson2314** [comment]
Isn't metaprogramming in general user-defined syntactic extension of the original language?

**Comment by gridaphobe** [comment]
No, metaprogramming is more general than that. For instance, in multi-stage programming the goal is strictly to produce more efficient code, specialized to a particular call-site. The classic example here is implementing the power function as a multi-stage program, where the compile-time stage unrolls the exponentiation into a sequence of multiplications.

In contrast, QuasiQuotes really are about adding new syntactic forms to the language, and I think having a little bit of extra syntax to highlight that helps. Yesterday at HIW someone asked about adding an extensible parser, or parser plugins, to GHC. I didn't make the connection then, but I think QuasiQuotes could be the solution.

**Comment by domenkozar** [comment]
It would be great to remove my only need for TH in my source by having multi-line string literals.

**Comment by Ericson2314** [comment]
I always want to steal the `QuasiQuotes` syntax for things like ideom brackets and https://github.com/conal/concat

**Comment by Ericson2314** [comment]
I would advocate that the way to do the things that @gridaphobe wants should not be the string-y, but a generalization of s-expressions. This what Rust did with its "token trees", we could have very similar "token trees" too, except they would use Haskell tokens (of course), and hopefully also support layout!

**Comment by brandonchinn178** [comment]
@gridaphobe actually, can we reopen this issue? I think there's a separate discussion for deprecating QuasiQuotes, even with native multiline string support

**Comment by gridaphobe** [comment]
Oh sure, I think that was just GitHub auto-closing due to the PR subject..
